core/src/main/java/org/apache/gravitino/cache/CaffeineMetaCache.java

```diff
@@ -0,0 +1,468 @@
+                CacheUtils.getIdFromEntity(entity),
+                identifier);
+          }
+
+          cacheIndex.remove(identifier.toString());
+        });
+  }
+
+  private void removeFromEntity(Entity rootEntity) {
+    NameIdentifier prefix = CacheUtils.getIdentFromEntity(rootEntity);
+
+    withLock(
+        () -> {
+          LOG.debug("Manual remove: prefix={}, removing matching entries", prefix);
+          // 1. find all keys starting with prefix
+          List<EntityKey> toRemovedId = Lists.newArrayList();
+          List<String> toRemovedIdent = Lists.newArrayList();
+
+          for (EntityKey idKey : cacheIndex.getValuesForKeysStartingWith(prefix.toString())) {
+            Entity entityToRemove = cacheData.getIfPresent(idKey);
+            toRemovedId.add(idKey);
+
+            if (entityToRemove != null) {
+              toRemovedIdent.add(CacheUtils.getIdentFromEntity(entityToRemove).toString());
+            }
+          }
+
+          if (toRemovedId.isEmpty()) {
+            return;
+          }
+
+          // 2, clean up id cache and name cache
+          cacheData.invalidateAll(toRemovedId);
+          // 3. remove from index tree
+          toRemovedIdent.forEach(ident -> cacheIndex.remove(ident));
+        });
+  }
+
+  private <KEY, VALUE> Caffeine<KEY, VALUE> newBaseBuilder(CacheConfig cacheConfig) {
+    Caffeine<Object, Object> builder = Caffeine.newBuilder();
+
+    if (cacheConfig.isWeightedCacheEnabled()) {
+      builder
+          .weigher(MetadataEntityWeigher.getInstance())
+          .maximumWeight(cacheConfig.getEntityCacheWeigherTarget());
+    } else {
+      builder.maximumSize(cacheConfig.getMaxSize());
+    }
+
+    if (cacheConfig.isExpirationEnabled()) {
+      builder.expireAfterWrite(
+          cacheConfig.getExpirationTime(), cacheConfig.getExpirationTimeUnit());
+    }
+
+    if (cacheConfig.isCacheStatusEnabled()) {
+      builder.recordStats();
+    }
+
+    return (Caffeine<KEY, VALUE>) builder;
+  }
+
+  private void initStatsScheduler() {
+    if (!cacheConfig.isCacheStatusEnabled()) {
+      return;
+    }
+
+    statsScheduler.scheduleAtFixedRate(
+        () -> {
+          try {
+            CacheStats idStats = cacheData.stats();
+            LOG.info(
+                "CacheStats [byId] hitRate={} missRate={} avgLoadTime={}ms",
+                idStats.hitRate(),
+                idStats.missRate(),
+                idStats.averageLoadPenalty() / 1_000_000.0);
+          } catch (Throwable t) {
+            LOG.warn("Error while logging cache stats", t);
+          }
+        },
+        1,
+        1,
+        TimeUnit.MINUTES);
+  }
+
+  private void withLock(Runnable action) {
+    opLock.lock();
+    try {
+      action.run();
+    } finally {
+      opLock.unlock();
+    }
+  }
+
+  /**
+   * Represents a key for the cache.
+   *
+   * <p>The key consists of the entity id and type.
+   */
+  static final class EntityKey {
```

Comment:
I think maybe the name `EntityIdAndType` is better than `EntityKey`?
Otherwise, maybe we need to move the `EntityKey` class to a standalone class, because another `XXXXMetaCache` depends on it.